Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.

While pandas adopt much code from NumPy, the difference is that Pandas is designed for tabular, heterogeneous data. NumPy, by difference, is best suited for working with homogeneous numerical array data.

The name Pandas is derived from the term 'panel data' (an econometrics term for multidimensional structured data sets).

Pandas has two data structures as follows:

1. A Series is a 1-dimensional labelled array that can hold data of any type (such as integer, string, boolean, float, python objects). Its axis labels are collectively called an index.
1. A DataFrame is a 2-dimensional labelled data structure with columns. It supports multiple data types.

In [1]:
# Import the numpy and pandas libraryimport numpy as np
import pandas as pd

In [2]:
#make series from list
data= [ 1,2,3,4,5]
series= pd.Series(data)
print(series)

0    1
1    2
2    3
3    4
4    5
dtype: int64


In [3]:
#make series from dictionary 
dic={'a':1,'b':2,'c':3}
se_dict=pd.Series(dic)
se_dict

a    1
b    2
c    3
dtype: int64

In [4]:
#create series with own data and index
data=[10,20,30]
index=[3,4,5]
s=pd.Series(data,index)
s

3    10
4    20
5    30
dtype: int64

In [5]:
#create dataframe from dictionary of list
data={
    'Name':['Arman','Arushi','urvi'],
    'Age':[23,34,25],
    'Salary':[20000,80000,30000]
    }
df_from_dict=pd.DataFrame(data)
df_from_dict

Name  Age  Salary
0   Arman   23   20000
1  Arushi   34   80000
2    urvi   25   30000

In [6]:
#create dataframe from list of dictionaries
data_dic=(
    {'Name':'Arman','Age':23,'Salary':20000},
    {'Name':'Anshuman','Age':25,'Salary':20000},
    {'Name':'Aman','Age':45,'Salary':20000},
    {'Name':'bharti','Age':34,'Salary':20000},
    {'Name':'urvi','Age':78,'Salary':195000}
)
df_from_listofdic=pd.DataFrame(data_dic)
df_from_listofdic

Name  Age  Salary
0     Arman   23   20000
1  Anshuman   25   20000
2      Aman   45   20000
3    bharti   34   20000
4      urvi   78  195000

In [7]:
import numpy as np
arrof_listofdict=np.array(data_dic)
arrof=np.array(df_from_listofdic)
arrof

array([['Arman', 23, 20000],
       ['Anshuman', 25, 20000],
       ['Aman', 45, 20000],
       ['bharti', 34, 20000],
       ['urvi', 78, 195000]], dtype=object)

In [57]:
#data from csv
data=pd.read_csv("graphicsAPIs.csv")
data

Manufacturer                     Device      CUDA  Metal    OpenCL  \
0          Nvidia        GeForce RTX 3090 Ti  260346.0    NaN  229738.0   
1          Nvidia             A100 80GB PCIe  259828.0    NaN  214586.0   
2          Nvidia             A100-PCIE-80GB  256292.0    NaN  207124.0   
3          Nvidia           GeForce RTX 3090  238123.0    NaN  204921.0   
4          Nvidia             A100-SXM4-40GB  237220.0    NaN  190489.0   
...           ...                        ...       ...    ...       ...   
1208          AMD       Vega 20 [Radeon VII]       NaN    NaN   96073.0   
1209        Other                VeriSilicon       NaN    NaN       NaN   
1210        Other  VideoCore VI HW (V3D-620)       NaN    NaN       NaN   
1211        Intel                Xe Graphics       NaN    NaN   49205.0   
1212        Intel      Xe Graphics (TGL GT2)       NaN    NaN       NaN   

        Vulkan  
0     141134.0  
1          NaN  
2          NaN  
3     138859.0  
4          NaN  
...        ...  
1208       NaN  
1209      73.0  
1210      76.0  
1211       NaN  
1212    8895.0  

[1213 rows x 6 columns]

In [59]:
df_from_dict['Name']

0     Arman
1    Arushi
2      urvi
Name: Name, dtype: object

1. Python loc() function can accept the boolean data.
2. The loc() function is label based data selecting method which means that we have to pass the name of the row or column which we want to select.
3. when we use :
1. Selecting Data According to Some Conditions
2. Selecting a Range of Rows From the DataFrame
3. Updating the Value of Any Column

In [75]:
df_from_dict.loc[0]#loc=row index

Name      Arman
Age          23
Salary    20000
Name: 0, dtype: object

when we use iloc:
1. Selecting Rows Using Integer Indices
2. Selecting a Range of Columns and Rows Simultaneously


In [76]:
df_from_dict.iloc[[0,2]] #iloc = column index

Name  Age  Salary
0  Arman   23   20000
2   urvi   25   30000

In [79]:
#2
df_from_dict.iloc[:2,:2] #iloc[row index start:end, col.index start:end]

Name  Age
0   Arman   23
1  Arushi   34

In [73]:
#accesing the specified element 
df_from_dict.at[1,'Age']

34

In [86]:
#iat 
df_from_dict
#df_from_dict.iat[2,2]

Name  Age  Salary
0   Arman   23   40000
1  Arushi   34   80000
2    urvi   25   60000

Data Manipulation with pandas

In [84]:
df_from_dict['Salary']=[40000,80000,60000]
df_from_dict

Name  Age  Salary
0   Arman   23   40000
1  Arushi   34   80000
2    urvi   25   60000

In [91]:
#creating a new column 
df_from_dict['City']=['Ratlam','Bhopal','Pune']
df_from_dict

Name  Age  Salary    City
0   Arman   23   40000  Ratlam
1  Arushi   34   80000  Bhopal
2    urvi   25   60000    Pune

In [ ]:
#to remove a column
df_from_dict.drop('City',axis=1,inplace=True)#it will check in column if axis =1 and then drop
#inplace=True means it will permanantly delete the column from df

Name  Age  Salary
0   Arman   23   40000
1  Arushi   34   80000
2    urvi   25   60000

In [95]:
#delete  a row from df
#df_from_dict.drop(1,inplace=True)
df_from_dict

Name  Age  Salary    City
0  Arman   23   40000  Ratlam
2   urvi   25   60000    Pune

In [96]:
#statistical summary 
df_from_dict.describe()

Age        Salary
count   2.000000      2.000000
mean   24.000000  50000.000000
std     1.414214  14142.135624
min    23.000000  40000.000000
25%    23.500000  45000.000000
50%    24.000000  50000.000000
75%    24.500000  55000.000000
max    25.000000  60000.000000

# Series

Pandas Series is a one-dimensional labelled array capable of holding any data type. However, a series is a sequence of similar data types, similar to an array, list, or column in a table.

It will assign a labelled index to each item in the pd.Series. By default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.

### Creating a Series

**To create a numeric series** 

In [9]:
# create a numeric series
num = range(1, 20)
pd.Series(num)

0      1
1      2
2      3
3      4
4      5
5      6
6      7
7      8
8      9
9     10
10    11
11    12
12    13
13    14
14    15
15    16
16    17
17    18
18    19
dtype: int64

The output also gives the data type of the series as `int64`

**To create an object series** 

In [10]:
# create an object series
string = "Hi" , "How" ,"are", "you", "?"
pd.Series(string)

0     Hi
1    How
2    are
3    you
4      ?
dtype: object

The output gives the data type of the series as `object`

**To create a series by giving both numeric and string values** 

In [11]:
# create a Series with an arbitrary list
mix = [345, 'London', 34.5, -34.45, 'Happy Birthday']
pd.Series(mix)

0               345
1            London
2              34.5
3            -34.45
4    Happy Birthday
dtype: object

Here the numeric values are treated as object.

**To set index values for a series**

In [12]:
# declare a list of marks
marks = [60, 89, 74, 86]

# declare a list of subjects
subject = ["Maths", "Science", "English" , "Social Science"]

marks_series = pd.Series(marks)
marks_series

0    60
1    89
2    74
3    86
dtype: int64

In [13]:
# create a series from the above list of marks with subjects as its index names
# index: adds the index 
marks_series = pd.Series(marks, index = subject)
marks_series

Maths             60
Science           89
English           74
Social Science    86
dtype: int64

The index is added using the argument `index=`. The data type of the series continues to be numeric.

**To print the values and index of the Series**

In [14]:
# print the index of the series
marks_series.index

Index(['Maths', 'Science', 'English', 'Social Science'], dtype='object')

In [15]:
# prints the values of the series
marks_series.values

array([60, 89, 74, 86], dtype=int64)

**To create a series from a dictionary**

In [16]:
# declare a dictionary
data = {'Maths': 60, 'Science': 89, 'English': 76, 'Social Science': 86}

# create a series from the dictionary
# the dictionary keys are the index names
# the dictionary values are the series values
pd.Series(data)

Maths             60
Science           89
English           76
Social Science    86
dtype: int64

On passing a `dict`, the index in the resulting Series will have the dict’s keys in given order.

**A series with missing values**

If we pass a key that is not defined, then its value will be `NaN`.

In [17]:
# create a list of subjects
subjects = ["Maths", "Science", "Art and Craft" , "Social Science"]

# declare a dictionary
data = {'Maths': 60, 'Science': 89, 'English': 76, 'Social Science': 86}

# create a series from the dictionary
marks_series = pd.Series(data, index = subjects)
marks_series

Maths             60.0
Science           89.0
Art and Craft      NaN
Social Science    86.0
dtype: float64

### Manipulating Series 

**To check for null values using `.isnull`**

In [18]:
# check for nulls in the data
marks_series.isnull()

Maths             False
Science           False
Art and Craft      True
Social Science    False
dtype: bool

`False` indicates that the value is not null.

**To check for null values using `.notnull`**

In [19]:
# check for nulls in the data
marks_series.notnull()

Maths              True
Science            True
Art and Craft     False
Social Science     True
dtype: bool

` True` indicates that the value is not null.

**To know the subjects in which marks score is more than 75**

In [20]:
# check for marks more than 75
marks_series[marks_series > 75]

Science           89.0
Social Science    86.0
dtype: float64

**To assign 68 marks to 'Art and Craft'**

In [21]:
marks_series

Maths             60.0
Science           89.0
Art and Craft      NaN
Social Science    86.0
dtype: float64

In [22]:
# assign 68 marks to 'Art and Craft'
marks_series['Art and Craft'] = 68
marks_series

Maths             60.0
Science           89.0
Art and Craft     68.0
Social Science    86.0
dtype: float64

**To check whether Maths marks are 73**

In [23]:
# check whether Maths marks are 73
marks_series.Maths == 73

False

In [24]:
# or you may use
marks_series['Maths'] == 73

False

**To create a series by generating numpy random numbers**

In [25]:
# create the numbers series 
# generate a sequence of 20 random numbers using randint()
num = np.random.randint(1, 20, 20)
num = pd.Series(num)
num

0     17
1      2
2      1
3     16
4     10
5     14
6     11
7     10
8      1
9      3
10     5
11    19
12     2
13    11
14     1
15    14
16     2
17    12
18    18
19     6
dtype: int32

In [26]:
(num ** 2) + 5

0     294
1       9
2       6
3     261
4     105
5     201
6     126
7     105
8       6
9      14
10     30
11    366
12      9
13    126
14      6
15    201
16      9
17    149
18    329
19     41
dtype: int32

**To find the square of the numbers series**

In [27]:
# declare a variable square 
# variable 'square' contains the 
square = pd.Series(num ** 2)
square

0     289
1       4
2       1
3     256
4     100
5     196
6     121
7     100
8       1
9       9
10     25
11    361
12      4
13    121
14      1
15    196
16      4
17    144
18    324
19     36
dtype: int32

In [28]:
square.index = num
square

17    289
2       4
1       1
16    256
10    100
14    196
11    121
10    100
1       1
3       9
5      25
19    361
2       4
11    121
1       1
14    196
2       4
12    144
18    324
6      36
dtype: int32

**Add a number 5 to every element of the series**

In [29]:
# add 5 to each element
square + 5

17    294
2       9
1       6
16    261
10    105
14    201
11    126
10    105
1       6
3      14
5      30
19    366
2       9
11    126
1       6
14    201
2       9
12    149
18    329
6      41
dtype: int32

The number 5 has been added to all the variables.

**To extract a value specifying the index**

In [30]:
# obtain the value having index 5
square[5]

25

All values with index as '14' are obtained

**To extract a range of values specifying the location**

In [31]:
# extract values having the location 
# obtain the values from the 3rd position till the 6th position
# indexing starts from 0
# the 7th position is not included
square[3: 7]

16    256
10    100
14    196
11    121
dtype: int32

**Usage of `.iloc`**

We use `.iloc` to get the values of the specified index of numbers 

In [32]:
# obtain the value in the 5th position
square.iloc[4]

100

In [33]:
# obtain the values from the 3rd till the 8th position
# indexing starts from 0
square.iloc[3:9]

16    256
10    100
14    196
11    121
10    100
1       1
dtype: int32

**Sorting a numeric series**

In [34]:
# create a pandas series
age = pd.Series([23, 45, np.nan, 41, 23, 34, 55, np.nan, 34, 20])

# print values
age

0    23.0
1    45.0
2     NaN
3    41.0
4    23.0
5    34.0
6    55.0
7     NaN
8    34.0
9    20.0
dtype: float64

In [35]:
# ascending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
age.sort_values(ascending = True)

9    20.0
0    23.0
4    23.0
5    34.0
8    34.0
3    41.0
1    45.0
6    55.0
2     NaN
7     NaN
dtype: float64

In [36]:
age.sort_values()

9    20.0
0    23.0
4    23.0
5    34.0
8    34.0
3    41.0
1    45.0
6    55.0
2     NaN
7     NaN
dtype: float64

In [37]:
# arrange in descending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in descending order
age.sort_values(ascending = False)

6    55.0
1    45.0
3    41.0
5    34.0
8    34.0
0    23.0
4    23.0
9    20.0
2     NaN
7     NaN
dtype: float64

**Sorting a categorical series**

In [38]:
# create a pandas series
string_values = pd.Series(["a", "j", "d", "f", "t", "a"])

# print the values
string_values

0    a
1    j
2    d
3    f
4    t
5    a
dtype: object

In [39]:
# ascending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)


In [40]:
# descending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in descending order


**Sorting based on index**

In [41]:
# recall the series 'square'
square

17    289
2       4
1       1
16    256
10    100
14    196
11    121
10    100
1       1
3       9
5      25
19    361
2       4
11    121
1       1
14    196
2       4
12    144
18    324
6      36
dtype: int32

In [42]:
# sort in ascending order based on index
# sort_index: sorts the series based on the index
# ascending : if specified as True, it sorts values in ascending order (default value is True)
square.sort_index(ascending = True)

1       1
1       1
1       1
2       4
2       4
2       4
3       9
5      25
6      36
10    100
10    100
11    121
11    121
12    144
14    196
14    196
16    256
17    289
18    324
19    361
dtype: int32

In [43]:
# sort in descending order based on index
# sort_index: sorts the series based on the index
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in descending order


**Rank a Series**

In [44]:
# recall the marks_series
marks_series

Maths             60.0
Science           89.0
Art and Craft     68.0
Social Science    86.0
dtype: float64

In [45]:
# rank the marks in ascending order
# rank(): ranks the values of a series 
marks_series.rank()

Maths             1.0
Science           4.0
Art and Craft     2.0
Social Science    3.0
dtype: float64

In [46]:
age.sort_values()

9    20.0
0    23.0
4    23.0
5    34.0
8    34.0
3    41.0
1    45.0
6    55.0
2     NaN
7     NaN
dtype: float64

In [47]:
age.sort_values().rank()

9    1.0
0    2.5
4    2.5
5    4.5
8    4.5
3    6.0
1    7.0
6    8.0
2    NaN
7    NaN
dtype: float64

In [48]:
age.sort_values().rank(method = 'dense')

9    1.0
0    2.0
4    2.0
5    3.0
8    3.0
3    4.0
1    5.0
6    6.0
2    NaN
7    NaN
dtype: float64

In [49]:
# rank the marks in descending order
# rank(): ranks the values of a series 
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in descending order


# DataFrames

A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (such as numeric, string, boolean).

The DataFrame has both row and column index; it can be thought of as a dict of Series all sharing the same index. In a data frame, the data is stored as one or more two-dimensional blocks rather than a list, dict, or some other collection of one-dimensional arrays.

While a DataFrame is physically two-dimensional, it can be used to represent higher dimensional data in a tabular format using hierarchical indexing.

### Creating DataFrames

**1. Creating a data frame a dictionary**

In [50]:
# create a dictionary
data = {'Subject': ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'],
        'Marks': (45, 65, 78, 65, 80, 78),
        'GPA': [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]}

# create the dataframe using DataFrame()
df_data = pd.DataFrame(data)
df_data

Subject  Marks  GPA
0     Maths     45  2.5
1   History     65  3.0
2   Science     78  3.5
3   English     65  2.0
4  Georaphy     80  4.0
5       Art     78  4.0

**To create dataframe from lists**

In [51]:
# declare the list 'Subject'
Subject = ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art']

# declare the list 'Marks'
Marks = [45, 65, 78, 65, 80, 78]

# declare the list 'CPA'
GPA = [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]

# create a DataFrame from the lists
# index: specifies the index names
df_marks = pd.DataFrame([Subject, Marks, GPA], index = ['Subject','Marks','GPA'])

# print the DataFrame
df_marks

0        1        2        3         4    5
Subject  Maths  History  Science  English  Georaphy  Art
Marks       45       65       78       65        80   78
GPA        2.5      3.0      3.5      2.0       4.0  4.0

However to want a vertical dataframe so we use `.T`. The 'T' stands for transpose.

In [52]:
# transpose the DataFrame
df_marks.T

Subject Marks  GPA
0     Maths    45  2.5
1   History    65  3.0
2   Science    78  3.5
3   English    65  2.0
4  Georaphy    80  4.0
5       Art    78  4.0

**To create dataframe from series**

In [53]:
# declare the series 'Subject'
Subject = pd.Series(['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'])

# declare the series 'Marks'
Marks = pd.Series([45, 65, 78, 65, 80, 78])

# declare the series 'GPA'
GPA = pd.Series([2.5, 3.0, 3.5, 2.0, 4.0, 4.0])

# create a DataFrame from the Series
# index: specifies the index names
df_marks = pd.DataFrame([Subject, Marks, GPA], index = ['Subject','Marks','GPA'])

# print the DataFrame
df_marks

0        1        2        3         4    5
Subject  Maths  History  Science  English  Georaphy  Art
Marks       45       65       78       65        80   78
GPA        2.5      3.0      3.5      2.0       4.0  4.0

However to want a vertical dataframe so we use `.T`. The 'T' stands for transpose.

In [54]:
# transpose the DataFrame
df_marks.T

Subject Marks  GPA
0     Maths    45  2.5
1   History    65  3.0
2   Science    78  3.5
3   English    65  2.0
4  Georaphy    80  4.0
5       Art    78  4.0

In [55]:
df_data['Marks into GPA'] = df_data['Marks'] * df_data['GPA']
df_data

Subject  Marks  GPA  Marks into GPA
0     Maths     45  2.5           112.5
1   History     65  3.0           195.0
2   Science     78  3.5           273.0
3   English     65  2.0           130.0
4  Georaphy     80  4.0           320.0
5       Art     78  4.0           312.0

**Remark:** Assign a name to the data frame and then use `.T` to transpose it.

**To read data from csv file**

In [56]:
df_basic_info = pd.read_csv("D:/Samatrix/Python/Notes/Pandas/basic_info.csv")

FileNotFoundError: [Errno 2] No such file or directory: 'D:/Samatrix/Python/Notes/Pandas/basic_info.csv'

In [62]:
df_basic_info

Age  Weight (in kg)  Height (in m)
0    45              60           1.35
1    12              43           1.21
2    54              78           1.50
3    26              65           1.21
4    68              50           1.32
5    21              43           1.52
6    10              32           1.65
7    57              34           1.61
8    75              23           1.24
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

In [16]:
type(df_basic_info)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

In [17]:
print(df_basic_info)

    Age  Weight (in kg)  Height (in m)
0    45              60           1.35
1    12              43           1.21
2    54              78           1.50
3    26              65           1.21
4    68              50           1.32
5    21              43           1.52
6    10              32           1.65
7    57              34           1.61
8    75              23           1.24
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85


**To print head of the data**

In [21]:
# display the first 5 rows of the DataFrame
# head(): displays the first 5 rows
# to display more rows, for example 15, use head(15)
# the default value is 5
df_basic_info.head()

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the first 9 rows, we write the number 9 in the parentheses.

In [19]:
# display 9 rows
df_basic_info.head(9)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32
5   21              43           1.52
6   10              32           1.65
7   57              34           1.61
8   75              23           1.24

**To print tail of the data**

In [23]:
# to display the last 5 rows
df_basic_info.tail()

Age  Weight (in kg)  Height (in m)
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the last 14 rows, we write the number 14 in the parentheses.

In [24]:
# to display the last 14 rows
df_basic_info.tail(14)

Age  Weight (in kg)  Height (in m)
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

**To obtain the dimension of the data**

In [25]:
# to display the shape of the data
df_basic_info.shape

(23, 3)

**To know the data types of a data frame**

In [28]:
# to know the type of each variable
df_basic_info.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

We see the data type of each variable.

**To know some information of the data**

In [29]:
# to know information on the variables in the data
df_basic_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`. 

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.



**To check the data type of column in the data frame**

In [ ]:
# check the type of each variable
type(df_basic_info.Age)

In [ ]:
# check the type of each variable
type(df_basic_info["Weight (in kg)"])

In [ ]:
# check the type of each variable
type(df_basic_info["Height (in m)"])

**Note that every column of the data frame is a pandas Series.**

## Manipulating DataFrames 

#### Add new column and rows

**Remark:**   
1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
2. New columns cannot be created with the ` df_basic_info.BMI ` syntax.
                   



**Adding a new column to the data set**

In [64]:
# create a new variable BMI
df_basic_info['BMI'] = df_basic_info['Weight (in kg)'] / (df_basic_info['Height (in m)'] ** 2)
df_basic_info.head()

Age  Weight (in kg)  Height (in m)        BMI
0   45              60           1.35  32.921811
1   12              43           1.21  29.369579
2   54              78           1.50  34.666667
3   26              65           1.21  44.395875
4   68              50           1.32  28.696051

In [65]:
# check the shape of the DataFrame
df_basic_info.shape

(23, 4)

**Adding a new row to the data set**

A new row can be added using the `loc`

In [66]:
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI
0    45              60           1.35  32.921811
1    12              43           1.21  29.369579
2    54              78           1.50  34.666667
3    26              65           1.21  44.395875
4    68              50           1.32  28.696051
5    21              43           1.52  18.611496
6    10              32           1.65  11.753903
7    57              34           1.61  13.116778
8    75              23           1.24  14.958377
9    32              21           1.52   9.089335
10   23              53           1.50  23.555556
11   34              65           1.76  20.983988
12   55              89           1.65  32.690542
13   23              45           1.75  14.693878
14   56              76           1.69  26.609713
15   67              78           1.85  22.790358
16   26              65           1.21  44.395875
17   56              74           1.69  25.909457
18   67              78           1.85  22.790358
19   26              65           1.21  44.395875
20   68              50           1.32  28.696051
21   56              76           1.69  26.609713
22   67              78           1.85  22.790358

In [67]:
# add a new row at the end of the DataFrame
df_basic_info.loc[23] = [45, 85, 1.8, 26.3]

# display the DataFrame
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0            43.0           1.52  18.611496
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0            53.0           1.50  23.555556
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
23  45.0            85.0           1.80  26.300000

We see that a new column number 23 has be added to the data.

**Indexing a dataframe using `.iloc`**

`DataFrame.iloc[]` method is used when the index label of a data frame is something other than numeric series of 0, 1, 2, 3….n or in case the user doesn’t know the index label. 

We shall work on the BMI data set.

**Select the 2nd row**

In [68]:
# select the second row
df_basic_info.iloc[1]

Age               12.000000
Weight (in kg)    43.000000
Height (in m)      1.210000
BMI               29.369579
Name: 1, dtype: float64

**Select fourth, seventh and tenth rows**

In [69]:
# select the 4th, 7th and the 10th row
x = [4,7,10]
df_basic_info.iloc[x]

Age  Weight (in kg)  Height (in m)        BMI
4   68.0            50.0           1.32  28.696051
7   57.0            34.0           1.61  13.116778
10  23.0            53.0           1.50  23.555556

We use two square brackets since we are passing a list of row numbers to be accessed.

**Select 12th to 17th rows**

In [70]:
# select rows 
df_basic_info.iloc[12: 18]

Age  Weight (in kg)  Height (in m)        BMI
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457

**Select the 1st column**

In [71]:
# select the 1st column
df_basic_info.iloc[0: 30, 0: 3]

Age  Weight (in kg)  Height (in m)
0   45.0            60.0           1.35
1   12.0            43.0           1.21
2   54.0            78.0           1.50
3   26.0            65.0           1.21
4   68.0            50.0           1.32
5   21.0            43.0           1.52
6   10.0            32.0           1.65
7   57.0            34.0           1.61
8   75.0            23.0           1.24
9   32.0            21.0           1.52
10  23.0            53.0           1.50
11  34.0            65.0           1.76
12  55.0            89.0           1.65
13  23.0            45.0           1.75
14  56.0            76.0           1.69
15  67.0            78.0           1.85
16  26.0            65.0           1.21
17  56.0            74.0           1.69
18  67.0            78.0           1.85
19  26.0            65.0           1.21
20  68.0            50.0           1.32
21  56.0            76.0           1.69
22  67.0            78.0           1.85
23  45.0            85.0           1.80

**Select the last column**

In [72]:
# select the last column
# the colon indicates that all the rows are selected
# -1 indicated that the last column is selected
df_basic_info.iloc[:, -1]

0     32.921811
1     29.369579
2     34.666667
3     44.395875
4     28.696051
5     18.611496
6     11.753903
7     13.116778
8     14.958377
9      9.089335
10    23.555556
11    20.983988
12    32.690542
13    14.693878
14    26.609713
15    22.790358
16    44.395875
17    25.909457
18    22.790358
19    44.395875
20    28.696051
21    26.609713
22    22.790358
23    26.300000
Name: BMI, dtype: float64

To select the last column we use -1, to select the second last column we use -2

**Select the first two columns**

In [73]:
# select the 1st and 2nd columns
# the colon indicates that all the rows are selected
# 0:2 indicates that the 1st and 2nd columns are selected
df_basic_info.iloc[:, 0:2]

Age  Weight (in kg)
0   45.0            60.0
1   12.0            43.0
2   54.0            78.0
3   26.0            65.0
4   68.0            50.0
5   21.0            43.0
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0            53.0
11  34.0            65.0
12  55.0            89.0
13  23.0            45.0
14  56.0            76.0
15  67.0            78.0
16  26.0            65.0
17  56.0            74.0
18  67.0            78.0
19  26.0            65.0
20  68.0            50.0
21  56.0            76.0
22  67.0            78.0
23  45.0            85.0

**Select the first two columns and 5 to 10 rows**

In [74]:
# 5:11 indicates that the 5th to 10th rows will be selected
# 0:2 indicates that the 1st and 2nd columns be selected
df_basic_info.iloc[5: 11, 0: 2]

Age  Weight (in kg)
5   21.0            43.0
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0            53.0

**Indexing a dataframe using `.loc`**

`DataFrame.loc[]` method is a method that takes only index labels and returns row or dataframe if the index label exists in the caller data frame. <br>
`DataFrame.loc[Row_names, column_names]` is used to select or index rows or columns based on their name.

**Select 1 to 5 rows and 2nd and 4th columns**

In [75]:
# 1:5 indicates that rows with indices 1, 2, 3, 4 and 5 are selected
# ["Weight (in kg)","BMI"] indicates that the specified columns be selected
df_basic_info.loc[1:5, ["Weight (in kg)", "BMI"]]

Weight (in kg)        BMI
1            43.0  29.369579
2            78.0  34.666667
3            65.0  44.395875
4            50.0  28.696051
5            43.0  18.611496

**Note:** the row names are numbers 

**Selecting columns by specifying column names**

**Select the column 'Age'**

In [76]:
# select the coumn 'Age'
df_basic_info.Age

0     45.0
1     12.0
2     54.0
3     26.0
4     68.0
5     21.0
6     10.0
7     57.0
8     75.0
9     32.0
10    23.0
11    34.0
12    55.0
13    23.0
14    56.0
15    67.0
16    26.0
17    56.0
18    67.0
19    26.0
20    68.0
21    56.0
22    67.0
23    45.0
Name: Age, dtype: float64

**Remark:** Using this method we can select only one column.

In [77]:
# OR
df_basic_info['Age']

0     45.0
1     12.0
2     54.0
3     26.0
4     68.0
5     21.0
6     10.0
7     57.0
8     75.0
9     32.0
10    23.0
11    34.0
12    55.0
13    23.0
14    56.0
15    67.0
16    26.0
17    56.0
18    67.0
19    26.0
20    68.0
21    56.0
22    67.0
23    45.0
Name: Age, dtype: float64

**Select the column 'Age' and 'BMI'**

In [78]:
# select two columns
# the column names are passed in a list
df_basic_info[['Age', 'BMI']]

Age        BMI
0   45.0  32.921811
1   12.0  29.369579
2   54.0  34.666667
3   26.0  44.395875
4   68.0  28.696051
5   21.0  18.611496
6   10.0  11.753903
7   57.0  13.116778
8   75.0  14.958377
9   32.0   9.089335
10  23.0  23.555556
11  34.0  20.983988
12  55.0  32.690542
13  23.0  14.693878
14  56.0  26.609713
15  67.0  22.790358
16  26.0  44.395875
17  56.0  25.909457
18  67.0  22.790358
19  26.0  44.395875
20  68.0  28.696051
21  56.0  26.609713
22  67.0  22.790358
23  45.0  26.300000

**Conditional subsetting**

Selecting rows where the value of `Age` is greater than 47 

In [79]:
# to selct rows where the Age is greater than 47
df_basic_info[df_basic_info['Age'] > 47]

Age  Weight (in kg)  Height (in m)        BMI
2   54.0            78.0           1.50  34.666667
4   68.0            50.0           1.32  28.696051
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
12  55.0            89.0           1.65  32.690542
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358

Subsetting the age more than 40 or weight column value more than 65

In [80]:
# to select rows where either age is greater than 40 or weight is more than 65
df_basic_info[(df_basic_info['Age'] > 40) & (df_basic_info['Weight (in kg)'] > 65)]

Age  Weight (in kg)  Height (in m)        BMI
2   54.0            78.0           1.50  34.666667
12  55.0            89.0           1.65  32.690542
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
23  45.0            85.0           1.80  26.300000

Subsetting the age and weight column value more than 65


In [81]:
# select rows where both age and weight are more than 65
df_basic_info[(df_basic_info['Age'] > 65) & (df_basic_info['Weight (in kg)'] > 65)]

Age  Weight (in kg)  Height (in m)        BMI
15  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
22  67.0            78.0           1.85  22.790358

**Sort the data frame on the basis of values in a column**

Each column of a pandas DataFrame is treated as a pandas Series. The `.sort_values()` in DataFrames works similar to the `pandas.Series`.

In [82]:
# sort the data frame on basis of 'Age' values
# by default the values will get sorted in ascending order


In [83]:
# Note: 'ascending = False' will sort the data frame in descending order

**Rank the dataframe**

In [84]:
# rank the data frame 'data' in descending order based on 'BMI'
# 'method = min' assigns the minimum rank to highest equal value of 'BMI' 
df_basic_info['Rank_min'] = df_basic_info['BMI'].rank(method = 'min')
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI  Rank_min
0   45.0            60.0           1.35  32.921811      20.0
1   12.0            43.0           1.21  29.369579      18.0
2   54.0            78.0           1.50  34.666667      21.0
3   26.0            65.0           1.21  44.395875      22.0
4   68.0            50.0           1.32  28.696051      16.0
5   21.0            43.0           1.52  18.611496       6.0
6   10.0            32.0           1.65  11.753903       2.0
7   57.0            34.0           1.61  13.116778       3.0
8   75.0            23.0           1.24  14.958377       5.0
9   32.0            21.0           1.52   9.089335       1.0
10  23.0            53.0           1.50  23.555556      11.0
11  34.0            65.0           1.76  20.983988       7.0
12  55.0            89.0           1.65  32.690542      19.0
13  23.0            45.0           1.75  14.693878       4.0
14  56.0            76.0           1.69  26.609713      14.0
15  67.0            78.0           1.85  22.790358       8.0
16  26.0            65.0           1.21  44.395875      22.0
17  56.0            74.0           1.69  25.909457      12.0
18  67.0            78.0           1.85  22.790358       8.0
19  26.0            65.0           1.21  44.395875      22.0
20  68.0            50.0           1.32  28.696051      16.0
21  56.0            76.0           1.69  26.609713      14.0
22  67.0            78.0           1.85  22.790358       8.0
23  45.0            85.0           1.80  26.300000      13.0

From the above data frame, we can see that 'BMI = 44.395875' is repeating thrice; thus the method = 'min' will assign the minimum rank (=1) to all the three values of BMI. The rank '4' will be assigned to the second largest value of BMI and so on. Thus, there is no rank equal to 2 and 3.

In [ ]:
# method = 'dense' assigns same rank to all the same BMI values


Here, the dense method assigns minimum rank (=1) to the minimum value (=9.089335) of the BMI. Rank 2 will be assigned to next smallest value and so on. The value 44.395875 which repeats thrice has the same rank - 18 to the three observations.

**To check for missing values**

In [ ]:
# read the dataset
df_basic_info_missing = pd.read_csv("basic_info_missingdata.csv")

# print the dataset
df_basic_info_missing

Check for missing values

In [ ]:
# isnull() returns True for a missing value
# sum() gives the sum of True values


The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.